hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/StreamWriteFunction.java

```diff
@@ -463,16 +439,15 @@ private void registerMetrics() {
     writeMetrics.registerMetrics();
   }
 
-  protected List<WriteStatus> writeBucket(String instant, DataBucket bucket, List<HoodieRecord> records) {
-    bucket.preWrite(records);
+  protected List<WriteStatus> writeRecords(String instant, List<HoodieRecord> records) {
     writeMetrics.startFileFlush();
-    List<WriteStatus> statuses = writeFunction.apply(records, instant);
+    List<WriteStatus> statuses = writeFunction.apply(deduplicateRecordsIfNeeded(records), instant);
```

Comment:
The record loc of the first record to write must be ensured, how can we avoid the wrong loc if the first record has been dropped through deduplication.

Reply:
It's because all records in the bucket will have identical location. The reason is how we decide which `bucketID` to use for each record to buffer.

For each incoming record in `StreamWriteFunction` we call `bufferRecord(value)`. And first step in `bufferRecord(value)` is to call `getBucketID(value)`:
https://github.com/apache/hudi/blob/47e5dddf7e64f79728398233314faf72c02e5283/hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/StreamWriteFunction.java#L333-L336
which means that each record in the same bucket (with the same `bucketID`) has current location, and all records in the the same bucket will have the same `fileId` and `partitionPath`.

As a result, all records in the bucket will have the same current location, and we don't need to rewrite it for the first record in the bucket before flush even after deduplication.

Reply:
The reason for preWrite is to update the instant time of the first record as either `I` or `U`, so that the correspondent write handle can be choosed. Of course the records in the same bucket has the same loc, but that is not what we design the preWrite for.

Reply:
But `preWrite` doesn't update instant time, and uses instant time of some new first record after deduplication:
https://github.com/apache/hudi/blob/47e5dddf7e64f79728398233314faf72c02e5283/hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/StreamWriteFunction.java#L240-L247

We use `first = records.get(0)` for `preWrite`, which means, that `first` is **some random record after deduplication** if the initial first one has been removed.
And we use almost all for this random record as it is:
- `first.getKey()`
- `first.getData()`
- `first.getOperation()`
- **`first.getCurrentLocation().getInstantTime()`**  <--

including instant time. The only difference is `fileID`, which is the same for all record in `DataBucket`, because:
https://github.com/apache/hudi/blob/47e5dddf7e64f79728398233314faf72c02e5283/hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/util/StreamerUtil.java#L317-L318
`fileID` defines `bucketID`.

Reply:
yeah, I lost some memories for the deduplication, the instant can be ensured because we always choose the precedding one during the deduplication, and the `preWrite` can indeed be got rid of after #12054.